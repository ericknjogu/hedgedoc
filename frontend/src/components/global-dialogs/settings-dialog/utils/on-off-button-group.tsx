/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { testId } from '../../../../utils/test-id'
import { SettingsToggleButton } from './settings-toggle-button'
import React, { useCallback, useMemo } from 'react'
import { ToggleButtonGroup } from 'react-bootstrap'

enum OnOffState {
  ON,
  OFF
}

export interface OnOffButtonGroupProps {
  value: boolean
  onSelect: (value: boolean) => void
  name: string
  overrideButtonOnI18nKey?: string
  overrideButtonOffI18nKey?: string
}

/**
 * Shows a button group that is used to toggle a setting on or off.
 *
 * @param onSelect callback that is executed if the on/off value has changed
 * @param value the current on/off value that should be visible
 * @param name A unique name for the button group to allow the browser to distinguish between multiple ones
 * @param overrideButtonOnI18nKey Set to override the i18n key for the on-button
 * @param overrideButtonOffI18nKey Set to override the i18n key for the off-button
 */
export const OnOffButtonGroup: React.FC<OnOffButtonGroupProps> = ({
  onSelect,
  value,
  name,
  overrideButtonOffI18nKey,
  overrideButtonOnI18nKey
}) => {
  const buttonGroupValue = useMemo(() => (value ? OnOffState.ON : OnOffState.OFF), [value])
  const onButtonSelect = useCallback(
    (value: OnOffState) => {
      onSelect(value === OnOffState.ON)
    },
    [onSelect]
  )

  return (
    <ToggleButtonGroup type='radio' name={name} value={buttonGroupValue}>
      <SettingsToggleButton
        onSelect={onButtonSelect}
        selected={buttonGroupValue === OnOffState.ON}
        value={OnOffState.ON}
        i18nKeyTooltip={overrideButtonOnI18nKey ?? 'common.on'}
        i18nKeyLabel={overrideButtonOnI18nKey ?? 'common.on'}
        {...testId('onOffButtonGroupOn')}
      />
      <SettingsToggleButton
        onSelect={onButtonSelect}
        selected={buttonGroupValue === OnOffState.OFF}
        value={OnOffState.OFF}
        i18nKeyTooltip={overrideButtonOffI18nKey ?? 'common.off'}
        i18nKeyLabel={overrideButtonOffI18nKey ?? 'common.off'}
        {...testId('onOffButtonGroupOff')}
      />
    </ToggleButtonGroup>
  )
}
