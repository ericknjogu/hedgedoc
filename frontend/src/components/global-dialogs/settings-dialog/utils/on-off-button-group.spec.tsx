/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { mockI18n } from '../../../../test-utils/mock-i18n'
import { OnOffButtonGroup } from './on-off-button-group'
import { act, render, screen } from '@testing-library/react'

describe('Settings On-Off Button Group', () => {
  beforeAll(mockI18n)

  it('can switch value', async () => {
    let value = false
    const onSelect = (newValue: boolean) => (value = newValue)

    const view = render(<OnOffButtonGroup name={'test'} value={value} onSelect={onSelect} />)
    expect(view.container).toMatchSnapshot()
    const onButton = await screen.findByTestId('onOffButtonGroupOn')
    await act<void>(() => {
      onButton.click()
    })
    expect(value).toBeTruthy()

    view.rerender(<OnOffButtonGroup name={'test'} value={value} onSelect={onSelect} />)
    expect(view.container).toMatchSnapshot()
    const offButton = await screen.findByTestId('onOffButtonGroupOff')
    await act<void>(() => {
      offButton.click()
    })
    expect(value).toBeFalsy()
  })

  it('accepts custom labels', () => {
    const view = render(
      <OnOffButtonGroup
        name={'test'}
        value={true}
        onSelect={() => {}}
        overrideButtonOnI18nKey={'test.custom-on'}
        overrideButtonOffI18nKey={'test.custom-off'}
      />
    )
    expect(view.container).toMatchSnapshot()
  })
})
