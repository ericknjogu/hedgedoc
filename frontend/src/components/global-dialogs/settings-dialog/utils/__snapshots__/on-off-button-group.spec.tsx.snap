// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings On-Off Button Group accepts custom labels 1`] = `
<div>
  <div
    class="btn-group"
    role="group"
  >
    <button
      class="btn btn-secondary"
      data-testid="onOffButtonGroupOn"
      name="test"
      title="test.custom-on"
      type="radio"
    >
      test.custom-on
    </button>
    <button
      class="btn btn-outline-secondary"
      data-testid="onOffButtonGroupOff"
      name="test"
      title="test.custom-off"
      type="radio"
    >
      test.custom-off
    </button>
  </div>
</div>
`;

exports[`Settings On-Off Button Group can switch value 1`] = `
<div>
  <div
    class="btn-group"
    role="group"
  >
    <button
      class="btn btn-outline-secondary"
      data-testid="onOffButtonGroupOn"
      name="test"
      title="common.on"
      type="radio"
    >
      common.on
    </button>
    <button
      class="btn btn-secondary"
      data-testid="onOffButtonGroupOff"
      name="test"
      title="common.off"
      type="radio"
    >
      common.off
    </button>
  </div>
</div>
`;

exports[`Settings On-Off Button Group can switch value 2`] = `
<div>
  <div
    class="btn-group"
    role="group"
  >
    <button
      class="btn btn-secondary"
      data-testid="onOffButtonGroupOn"
      name="test"
      title="common.on"
      type="radio"
    >
      common.on
    </button>
    <button
      class="btn btn-outline-secondary"
      data-testid="onOffButtonGroupOff"
      name="test"
      title="common.off"
      type="radio"
    >
      common.off
    </button>
  </div>
</div>
`;
